src/Unleash/Internal/ToggleBootstrapHandler.cs

```diff
@@ -0,0 +1,37 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using Unleash.Serialization;
+
+namespace Unleash.Internal
+{
+    internal class ToggleBootstrapHandler : IToggleBootstrapHandler
+    {
+        private readonly IJsonSerializer jsonSerializer;
+        private readonly IFileSystem fileSystem;
+        private readonly IToggleBootstrapProvider toggleBootstrapProvider;
+        private readonly Encoding encoding;
+
+        public ToggleBootstrapHandler(UnleashSettings settings)
+        {
+            jsonSerializer = settings.JsonSerializer;
+            fileSystem = settings.FileSystem;
+            toggleBootstrapProvider = settings.ToggleBootstrapProvider;
+            encoding = settings.Encoding;
+        }
+
+        public ToggleCollection Read()
+        {
+            if (toggleBootstrapProvider == null)
+                return null;
+
+            var bootstrapContent = toggleBootstrapProvider.Read();
+            var bootstrapContentBuffer = encoding.GetBytes(bootstrapContent ?? "");
+            using (var bootstrapStream = new MemoryStream(bootstrapContentBuffer))
```

Comment:
The bootstrap provider seems to return a string, which means that it's already read a byte stream from a socket/filehandle and converted that to a string, do we have convert it back to byte stream?

Reply:
Yeah this is actually one of the things i've not been 100% happy with implementation wise, and want to revisit before completion, so i'm happy to get comments on this!

The interface used to deserialize JSON only takes a Stream input, so in order not to grow the scope too much, and be close to the java reference sdk, provide an easy interface for others to implement + skip some hoops in testing, this is the tradeoff i opted for to keep moving.

The two options as i see them are to convert back to exposing the original streams and expect the caller to dispose properly, or build an overload in the JSON deserialization abstraction and reflection code to handle a string parameter. I might be leaning ever so slightly in preference towards the latter for the reasons mentioned, but will happily take your thoughts on that!

Sorry for the long reply

Reply:
Ha! As I was looking at this I was thinking to myself "I bet this is because this interface is unwieldy"

So I don't have a clear, simple or obvious solution to this issue, but I do have some thoughts and I'm going to preface this by saying this is the Rust dev in me saying this, not the C# dev so if it makes you uncomfortable, just ignore me. Right now I feel like the ToggleBootstrapHandler doesn't do very much, it seems to just be responsible to returning a result of a IToggleBootstrapProvider and parsing that string to a ToggleCollection. That abstraction seems to be leaking a little for two reasons: 1) The ToggleBootstrapHandler knows about FileSystems (does it need to?) and 2) We have a dance where we're doing string/bytestream conversions. I don't think we need this amount of abstraction, I would just collapse the responsibility for calling a JSON parser into the BootStrapProviders (right now they don't actually have anything to do with Toggles, they read a file and a socket and they could do that job against any datasource really). This also means there's no real validation that what the BootstrapProvider returns will actually parse to a ToglleCollection, so this allows us to be suuuuper careful about error handling (yup, the Rust dev talking again). Right now there's a bunch of exceptions that can bubble from parsing a JSON bytestream and we're not doing anything with them

(I see your long reply and I raise you this one!)